src/protocol/mod.rs

```diff
@@ -0,0 +1,53 @@
+use crate::packet::{HandshakeControlInfo, ShakeType};
+use crate::ControlPacket;
+use std::net::SocketAddr;
+use std::time::Instant;
+
+pub mod sender;
+
+pub enum Handshake {
+    Caller,
+    Listener(ControlPacket),
+    Rendezvous(ControlPacket),
+}
+
+impl Handshake {
+    pub fn handle_handshake<E>(
+        &self,
+        from: SocketAddr,
+        handshake: HandshakeControlInfo,
+    ) -> Result<Option<(ControlPacket, SocketAddr)>, E> {
+        match (self, handshake.shake_type) {
+            (Handshake::Rendezvous(packet), ShakeType::Conclusion) => {
+                Ok(Some((packet.clone(), from)))
+            }
+            (Handshake::Listener(packet), _) => Ok(Some((packet.clone(), from))),
+            (Handshake::Caller, _) | (Handshake::Rendezvous(_), _) => Ok(None),
```

Comment:
Fair, though it's a objective trade off between being explicit and succinct, I also may have left it there because Rendezvous actually should be doing something else, but I neglected to add a TODO comment.